File: src/codes/ref-synergy/py/bridge-strategy.py
from abc import ABC, abstractmethod

class MessageStrategy(ABC):
    @abstractmethod
    def compose(self, user: str, content: str) -> str:
        pass

class PlainMessageStrategy(MessageStrategy):
    def compose(self, user: str, content: str) -> str:
        return f"{user}: {content}"

class HtmlMessageStrategy(MessageStrategy):
    def compose(self, user: str, content: str) -> str:
        return f"<p><strong>{user}</strong>: {content}</p>"

class Notifier(ABC):
    def __init__(self, strategy: MessageStrategy):
        self.strategy = strategy

    @abstractmethod
    def notify(self, user: str, content: str) -> None:
        pass

class EmailNotifier(Notifier):
    def notify(self, user: str, content: str) -> None:
        message = self.strategy.compose(user, content)
        print(f"[Email] Sending: {message}")

class SlackNotifier(Notifier):
    def notify(self, user: str, content: str) -> None:
        message = self.strategy.compose(user, content)
        print(f"[Slack] Sending: {message}")

# 使用例
notifier = SlackNotifier(HtmlMessageStrategy())
notifier.notify("Taro", "Server restarted.")

File: src/codes/ref-synergy/ts/bridge-strategy.ts
interface MessageStrategy {
  compose(user: string, content: string): string;
}

class PlainMessageStrategy implements MessageStrategy {
  compose(user: string, content: string): string {
    return `${user}: ${content}`;
  }
}

class HtmlMessageStrategy implements MessageStrategy {
  compose(user: string, content: string): string {
    return `<p><strong>${user}</strong>: ${content}</p>`;
  }
}

abstract class Notifier {
  constructor(protected strategy: MessageStrategy) {}

  abstract notify(user: string, content: string): void;
}

class EmailNotifier extends Notifier {
  notify(user: string, content: string): void {
    const message = this.strategy.compose(user, content);
    console.log(`[Email] Sending: ${message}`);
  }
}

class SlackNotifier extends Notifier {
  notify(user: string, content: string): void {
    const message = this.strategy.compose(user, content);
    console.log(`[Slack] Sending: ${message}`);
  }
}

// 使用例
const notifier = new SlackNotifier(new HtmlMessageStrategy());
notifier.notify("Taro", "Server restarted.");
